module/zfs/zfs_chksum.c

```diff
@@ -0,0 +1,316 @@
+	off += snprintf(buf + off, size - off, "%8llu",
+	    (u_longlong_t)cs->bs1k);
+	off += snprintf(buf + off, size - off, "%8llu",
+	    (u_longlong_t)cs->bs4k);
+	off += snprintf(buf + off, size - off, "%8llu",
+	    (u_longlong_t)cs->bs16k);
+	off += snprintf(buf + off, size - off, "%8llu",
+	    (u_longlong_t)cs->bs64k);
+	off += snprintf(buf + off, size - off, "%8llu",
+	    (u_longlong_t)cs->bs256k);
+	off += snprintf(buf + off, size - off, "%8llu",
+	    (u_longlong_t)cs->bs1m);
+	(void) snprintf(buf + off, size - off, "%8llu\n",
+	    (u_longlong_t)cs->bs4m);
+
+	return (0);
+}
+
+static void *
+chksum_stat_kstat_addr(kstat_t *ksp, loff_t n)
+{
+	if (n < chksum_stat_cnt)
+		ksp->ks_private = (void *)(chksum_stat_data + n);
+	else
+		ksp->ks_private = NULL;
+
+	return (ksp->ks_private);
+}
+
+static void
+chksum_run(chksum_stat_t *cs, abd_t *abd, void *ctx, int round,
+    uint64_t *result)
+{
+	hrtime_t start;
+	uint64_t run_bw, run_time_ns, run_count = 0, size = 0;
+	uint32_t l, loops = 0;
+	zio_cksum_t zcp;
+
+	switch (round) {
+	case 1: /* 1k */
+		size = 1<<10; loops = 128; break;
+	case 2: /* 2k */
+		size = 1<<12; loops = 64; break;
+	case 3: /* 4k */
+		size = 1<<14; loops = 32; break;
+	case 4: /* 16k */
+		size = 1<<16; loops = 16; break;
+	case 5: /* 256k */
+		size = 1<<18; loops = 8; break;
+	case 6: /* 1m */
+		size = 1<<20; loops = 4; break;
+	case 7: /* 4m */
+		size = 1<<22; loops = 1; break;
+	}
+
+	kpreempt_disable();
+	start = gethrtime();
+	do {
+		for (l = 0; l < loops; l++, run_count++)
+			cs->func(abd, size, ctx, &zcp);
+
+		run_time_ns = gethrtime() - start;
+	} while (run_time_ns < MSEC2NSEC(1));
+	kpreempt_enable();
+
+	run_bw = size * run_count * NANOSEC;
+	run_bw /= run_time_ns;	/* B/s */
+	*result = run_bw/1024/1024; /* MiB/s */
+}
+
+static void
+chksum_benchit(chksum_stat_t *cs)
+{
+	abd_t *abd;
+	void *ctx = 0;
+	void *salt = &cs->salt.zcs_bytes;
+
+	/* allocate test memory via default abd interface */
+	abd = abd_alloc_linear(1<<22, B_FALSE);
+	memset(salt, 0, sizeof (cs->salt.zcs_bytes));
+	if (cs->init) {
+		ctx = cs->init(&cs->salt);
+	}
+
+	chksum_run(cs, abd, ctx, 1, &cs->bs1k);
+	chksum_run(cs, abd, ctx, 2, &cs->bs4k);
+	chksum_run(cs, abd, ctx, 3, &cs->bs16k);
+	chksum_run(cs, abd, ctx, 4, &cs->bs64k);
+	chksum_run(cs, abd, ctx, 5, &cs->bs256k);
+	chksum_run(cs, abd, ctx, 6, &cs->bs1m);
+	chksum_run(cs, abd, ctx, 7, &cs->bs4m);
+
+	/* free up temp memory */
+	if (cs->free) {
+		cs->free(ctx);
+	}
+	abd_free(abd);
+}
+
+/*
+ * Initialize and benchmark all supported implementations.
+ */
+static void
+chksum_benchmark(void)
+{
+
+#ifndef _KERNEL
+	/* we need the benchmark only for the kernel module */
+	return;
+#endif
+
+	chksum_stat_t *cs;
+	int cbid = 0, id;
+	uint64_t max = 0;
+
+	/* space for the benchmark times */
+	chksum_stat_cnt = 4;
+	chksum_stat_cnt += blake3_get_impl_count();
+	chksum_stat_data = (chksum_stat_t *)kmem_zalloc(
+	    sizeof (chksum_stat_t) * chksum_stat_cnt, KM_SLEEP);
```

Comment:
To be clear, I think it would be great to include the fletcher benchmark results here.  I just wanted to make sure we don't end up in a situation where we're running it twice.  That said, I'm fine with doing that in this PR or leaving it as a follow up bit of work.

Reply:
Given that it's fletcher4, of all things, does it matter that much if it runs two benchmark sets on initial startup?

Reply:
It's certainly not the end of the world.  It was just something I noticed which I thought should be pretty straight forward to sort out.